Add blog post on PodSecurityPolicy historical context (#36021)
kubernetes/website: content/en/blog/_posts/2022-08-16-PSP-historical-context/index.md

---
layout: blog
title: "PodSecurityPolicy: The Historical Context"
date: 2022-08-16
slug: podsecuritypolicy-the-historical-context
---

> **Author:** Mahé Tardy (Quarkslab)

Reviewer: looks good

The PodSecurityPolicy (PSP) admission controller will be removed as of
Kubernetes v1.25. Its deprecation was announced and detailed in the blog post
[PodSecurityPolicy Deprecation: Past, Present, and Future](/blog/2021/04/06/podsecuritypolicy-deprecation-past-present-and-future/),
published for the Kubernetes v1.21 release.

> This article aims at providing historical context on the birth and evolution of
> PSP, to explain why the feature never made it to stable and had to be removed
> in favor of a new in-tree admission controller. Let's recap what
> PodSecurityPolicy is, why it was necessary, the historical context of its
> addition to Kubernetes, and finally, why it was removed and replaced by Pod
> Security admission control.

Reviewer 2: ```suggestion
This article aims to provide historical context on the birth and evolution of PodSecurityPolicy
(PSP), explain why the feature never made it to stable, and show why it was removed and replaced by Pod
Security admission control.
```
Are lines 18-20 redundant? If so, how about this proposed revision?

Author: Yes correct, these things were written separately and now that I put them just asides it's redundant!

It should be the "motivation" to read the article.

PodSecurityPolicy, like other specialized admission control plugins, provided
fine-grained permissions on specific fields concerning the pod security

> settings as a built-in policy API. It acknowledged the fact that cluster
> administrators and cluster users are, most of the time, not the same people, and

Reviewer 2: ```suggestion
settings as a built-in policy API. It acknowledged that cluster
administrators and cluster users are usually not the same people, and
```

> that creating workloads in the form of a Pod or any resource that will
> create a Pod should not equal being "root on the cluster". PodSecurityPolicy

Reviewer 2: ```suggestion
that creating workloads in the form of a Pod or any resource that
creates a Pod should not require being "root on the cluster". PodSecurityPolicy
```

Author: This suggestion will change the idea that I want to express: the problem I want to explain here is that "if you are able to create a workload in a cluster without admission control, you can be root on any node of the cluster, i.e. root on the cluster" and not "you should not require being root on the cluster to be able to create a workload".

Do you think my suggestion is unclear or confusing in the first place? We could explain it more clearly!

could also encourage best practices by configuring more secure defaults through
mutation and decoupling low-level Linux security decisions from the deployment
process.

## The birth of PodSecurityPolicy

PodSecurityPolicy originated from OpenShift's SecurityContextConstraints

> (SCC) that were in the very first release of the Red Hat container platform,

Author: I wanted to avoid the repetition but it's far better like this!

even before Kubernetes 1.0. PSP was a stripped-down version of the SCC.

<!-- not sure the next paragraph is relevant -->

The origin of the creation of PSP is difficult to track notably because it was
mainly added before Kubernetes Enhancements Proposal (KEP) process, when design
proposals were still a thing. Indeed, the archive of the final
[design proposal](https://github.com/kubernetes/design-proposals-archive/blob/main/auth/pod-security-policy.md)

> is still available. Nevertheless, a KEP issue was created after the first pull
> requests were merged, and is [KEP number five](https://github.com/kubernetes/enhancements/issues/5).

Reviewer 2: ```suggestion
is still available. Nevertheless, a [KEP issue number five](https://github.com/kubernetes/enhancements/issues/5) 
was created after the first pull requests were merged.
```

Before adding the first piece of code that created the PSP, two main pull
requests were merged into Kubernetes, [the security context subresource](https://github.com/kubernetes/kubernetes/pull/7343)
that defined the new fields on pods' containers and [the service account](https://github.com/kubernetes/kubernetes/pull/7101)

> feature. Then, Kubernetes 1.0 was released on 10 July 2015 without a mechanism
> to restrict the security context and sensitive options of workloads, apart from
> scdeny, an admission control that just prevents usage of some fields in the
> security context, which still exists today.

Reviewer 2: ```suggestion
feature. Then, apart from the scdeny package, Kubernetes 1.0 was released on 10 July 2015 without 
a mechanism to restrict the security context and sensitive options of workloads.
The scdeny package is still in use today and creates an admission controller that prevents the usage of some fields in the security context.
```

Author: You are perfectly correct, it's [technically a package](https://github.com/kubernetes/kubernetes/blob/master/plugin/pkg/admission/securitycontext/scdeny/admission.go) but it feels a bit weird to talk about an admission plugin as a package, it seems a bit too generic!
```suggestion
feature. Then, apart from the scdeny admission plugin, Kubernetes 1.0 was released on 10 July 2015 without 
a mechanism to restrict the security context and sensitive options of workloads.
The scdeny plugin is still in use today and creates an admission controller that prevents the usage of some fields in the security context.
```
What do you think of that?

Reviewer 2: Looks great to me.

The roots of the PodSecurityPolicy were added with
[the very first pull request on security policy](https://github.com/kubernetes/kubernetes/pull/7893),
which added the design proposal with the new PSP object, based on the SCC (Security Context Constraints). It
was a long discussion of nine months, with back and forth from OpenShift's SCC,
many rebases, and the rename to PodSecurityPolicy that finally made it to

> upstream Kubernetes in February 2016. Now that Pod Security Policy object
> was created, the next step was to add an admission controller that could enforce

Reviewer 2: ```suggestion
upstream Kubernetes in February 2016. Now that the Pod Security Policy object
had been created, the next step was to add an admission controller that could enforce
```

these policies. The first step was to add the admission
[without taking into account the users or groups](https://github.com/kubernetes/kubernetes/pull/7893#issuecomment-180410539).
A specific [issue to bring PodSecurityPolicy to a usable state](https://github.com/kubernetes/kubernetes/issues/23217)
was added to keep track of the progress and a first version of the admission
controller was merged in [pull request named PSP admission](https://github.com/kubernetes/kubernetes/pull/24600)
in May 2016. Then around two months later, Kubernetes 1.3 was released.

Here is a timeline that recaps the main pull requests of the birth of the Pod
Security Policy and its admission controller with 1.0 and 1.3 releases as
reference points.

![Timeline of the main Pod Security Policy pull requests - THIS IS TEMPORARY](https://i.imgur.com/rP5mRUW.png)

After that, the PSP admission controller was enhanced by adding what was initially

> left aside, [the authorization mechanism](https://github.com/kubernetes/kubernetes/pull/33080),

Reviewer 2: ```suggestion
left aside. [The authorization mechanism](https://github.com/kubernetes/kubernetes/pull/33080),
```

> merged in early November 2016, allowing administrators to use multiple policies

Reviewer 2: ```suggestion
merged in early November 2016 allowed administrators to use multiple policies
```

> in a cluster that grant different levels of access for different types of users.

Reviewer 2: ```suggestion
in a cluster to grant different levels of access for different types of users.
```

> And later, a [pull request](https://github.com/kubernetes/kubernetes/pull/52849)
> was merged in October 2017 to fix [a design issue](https://github.com/kubernetes/kubernetes/issues/36184)
> on ordering the PSP between mutating and alphabetical order and continue to
> build the PSP admission as we know it. After that, many improvements and fixes
> followed to build the Pod Security Policy feature of recent Kubernetes releases.

Reviewer 2: ```suggestion
Later, a [pull request](https://github.com/kubernetes/kubernetes/pull/52849)
merged in October 2017 fixed [a design issue](https://github.com/kubernetes/kubernetes/issues/36184)
on ordering the PSP between mutating and alphabetical order, and continued to
build the PSP admission as we know it. After that, many improvements and fixes
followed to build the Pod Security Policy feature of recent Kubernetes releases.
```

```
TODO: put the extended timeline here
```

## The rise of Pod Security Admission 

<!-- this list is mostly Tim Allclair slides -->

> Despite the crucial issue it was trying to solve, PSP presented some majors

Reviewer 2: ```suggestion
Despite the crucial issue it was trying to solve, PSP presented some major
```

flaws:

- **Flawed authorization model** - users can create a pod if they have the
  **use** verb on the PSP that allows that pod or the pod's service account has
  the **use** permission on the allowing PSP.

> - **Difficult to roll out** - PSP fail-closed, i.e. in the absence of policy,

Reviewer 2: ```suggestion
- **Difficult to roll out** - PSP fail-closed. That is, in the absence of a policy,
```

  composability with a weak prioritization model, leading to unexpected
  mutation priority. It made it really difficult to combine PSP with other
  third-party admission controllers.
- **Require security knowledge** - effective usage still requires an
  understanding of Linux security primitives. e.g. MustRunAsNonRoot +
  AllowPrivilegeEscalation.

<!-- past paragraph may be a bit too harsh -->

The experience of the PSP concluded that most users care for two or three
policies, which led to the creation of the [Pod Security Standards](/docs/concepts/security/pod-security-standards/),
that define three policies:
- **Privileged** --- unrestricted policy.
- **Baseline** --- minimally restrictive policy, allowing the default pod
  configuration.
- **Restricted** --- security best practice policy.

The replacement for the PSP, the new [Pod Security Admission](/docs/concepts/security/pod-security-admission/)
is an in-tree, stable for Kubernetes v1.25, admission plugin to enforce these
standards at the namespace level. It makes it easier to enforce basic pod
security without deep security knowledge. For more sophisticated use cases, you

> might need to use a third-party solution, that can be easily combined with Pod

Reviewer 2: ```suggestion
might need a third-party solution that can be easily combined with Pod
```

---
layout: blog
title: "PodSecurityPolicy: The Historical Context"

> date: 2022-08-23

Author: ```suggestion
date: 2022-08-23T14:00:00-0800
```

slug: podsecuritypolicy-the-historical-context
---

**Author:** Mahé Tardy (Quarkslab)

> The PodSecurityPolicy (PSP) admission controller will be removed as of

Reviewer 3: ```suggestion
The PodSecurityPolicy (PSP) admission controller has been removed, as of
```

the form of a Pod or any resource that will create a Pod should not equal being
"root on the cluster". It could also encourage best practices by configuring
more secure defaults through mutation and decoupling low-level Linux security
decisions from the deployment process.

## The birth of PodSecurityPolicy

PodSecurityPolicy originated from OpenShift's SecurityContextConstraints
(SCC) that were in the very first release of the Red Hat OpenShift Container Platform,
even before Kubernetes 1.0. PSP was a stripped-down version of the SCC.

The origin of the creation of PodSecurityPolicy is difficult to track, notably
because it was mainly added before Kubernetes Enhancements Proposal (KEP)
process, when design proposals were still a thing. Indeed, the archive of the final
[design proposal](https://github.com/kubernetes/design-proposals-archive/blob/main/auth/pod-security-policy.md)
is still available. Nevertheless, a [KEP issue number five](https://github.com/kubernetes/enhancements/issues/5) 
was created after the first pull requests were merged.

Before adding the first piece of code that created PSP, two main pull

> requests were merged into Kubernetes, [the security context subresource](https://github.com/kubernetes/kubernetes/pull/7343)
> that defined the new fields on pods' containers and [the service account](https://github.com/kubernetes/kubernetes/pull/7101)

Reviewer 3: ```suggestion
requests were merged into Kubernetes, a [`SecurityContext` subresource](https://github.com/kubernetes/kubernetes/pull/7343)
that defined new fields on pods' containers, and the first iteration of the [ServiceAccount](https://github.com/kubernetes/kubernetes/pull/7101)
```

> feature. Then, apart from the scdeny admission plugin, Kubernetes 1.0 was released on 10 July 2015 without 
> a mechanism to restrict the security context and sensitive options of workloads.

Reviewer 3: ```suggestion
API.

Kubernetes 1.0 was released on 10 July 2015 without any mechanism to restrict the
security context and sensitive options of workloads, other than an alpha-quality
SecurityContextDeny admission plugin (then known as `scdeny`).
```

> The scdeny plugin is still in use today and creates an admission controller that prevents the usage of some fields in the security context.

Reviewer 3: ```suggestion
The [SecurityContextDeny plugin](https://kubernetes.io/docs/reference/access-authn-authz/admission-controllers/#securitycontextdeny)
is still in Kubernetes today (as an alpha feature) and creates an admission controller that
prevents the usage of some fields in the security context.
```

  mutation priority. It made it really difficult to combine PSP with other
  third-party admission controllers.
- **Require security knowledge** - effective usage still requires an
  understanding of Linux security primitives. e.g. MustRunAsNonRoot +
  AllowPrivilegeEscalation.

The experience with PodSecurityPolicy concluded that most users care for two or three
policies, which led to the creation of the [Pod Security Standards](/docs/concepts/security/pod-security-standards/),
that define three policies:
- **Privileged** - unrestricted policy.
- **Baseline** - minimally restrictive policy, allowing the default pod
  configuration.
- **Restricted** - security best practice policy.

The replacement for PSP, the new [Pod Security Admission](/docs/concepts/security/pod-security-admission/)
is an in-tree, stable for Kubernetes v1.25, admission plugin to enforce these
standards at the namespace level. It makes it easier to enforce basic pod
security without deep security knowledge. For more sophisticated use cases, you
might need a third-party solution that can be easily combined with Pod Security
Admission.

## What's next

> For further details on the SIG auth process to PodSecurityPolicy removal and

Reviewer 3: ```suggestion
For further details on the SIG Auth processes, covering PodSecurityPolicy removal and
```

---
layout: blog
title: "PodSecurityPolicy: The Historical Context"

> date: 2022-08-16

Author: ```suggestion
date: 2022-08-23
```

---
layout: blog
title: "PodSecurityPolicy: The Historical Context"

> date: 2022-08-23T14:00:00-0800

Reviewer 3: ```suggestion
date: 2022-08-23T15:00:00-0800
```
Same UTC day, and after the Kubernetes release actually happened.
